Assemble.c

```diff
@@ -1975,6 +2007,15 @@ int assemble_container_content(struct supertype *st, int mdfd,
 	old_raid_disks = content->array.raid_disks - content->delta_disks;
 	avail = xcalloc(content->array.raid_disks, 1);
 	for (dev = content->devs; dev; dev = dev->next) {
+		// skip dev if event count too low
+		if ((int)dev->events < min_assembly_seq) {
```

Comment:
All this code looks great.  Can't find any way out of it by inspection that'd come back and bite us!  

I threw a bunch of tests at it, and only found one weird thing.  I think we might be OK with it, since heal is used to dealing with random shit, but just in case...

1. created and ran array, wrote to it to advance sequence number.
2. popped disk A out of target, wrote to array.  disk A is kicked out of the subarray.
3. wrote to array again for good measure.
4. stopped subarray
5. put disk A back into target. 
6. popped disk B out of its target
7. assemble and run the subarray.

Mdadm ends up bumping up against this check here, at line 2011 (I added a "XXX 2" to this message so I could see which one it hit.)  Then, it proceeded to assemble the subarray with no devices.  `/proc/mdstat` now looks like this:

```
Personalities : [raid1]
md1 : inactive super external:/md0/0

md0 : inactive sdc[0](S) sdd[1](S)
      65536 blocks super external:ddf

unused devices: <none>
```

Here's what mdadm logged:

```
[root@localhost mdadm]# ./mdadm --assemble --run --mdvote /dev/md1 /dev/md0 --bitmap /tmp/typically-terrible-mike-filename
mdadm: load_sys: /sys/block/md0/md/metadata_version: external:ddf
mdadm: load_super_ddf_all: load_super_ddf_all()
mdadm: load_sys: /sys/block/md0/md/metadata_version: external:ddf
mdadm: load_sys: /sys/block/md0/md/level:
mdadm: load_sys: /sys/block/md0/md/dev-sdc/slot: none
mdadm: load_sys: /sys/block/md0/md/dev-sdc/block/dev: 8:32
mdadm: load_sys: /sys/block/md0/md/dev-sdc/block/device/state: running
mdadm: load_sys: /sys/block/md0/md/dev-sdc/state: spare
mdadm: load_sys: /sys/block/md0/md/dev-sdd/slot: none
mdadm: load_sys: /sys/block/md0/md/dev-sdd/block/dev: 8:48
mdadm: load_sys: /sys/block/md0/md/dev-sdd/block/device/state: running
mdadm: load_sys: /sys/block/md0/md/dev-sdd/state: spare
mdadm: load_ddf_headers: metadata: cannot read anchor block: Input/output error
mdadm: load_ddf_headers: metadata: cannot read anchor block: Input/output error
mdadm: load_super_ddf_all: unable to read ddf local from device: 8:48
mdadm: mdvote_log_res: "GET /e63f1c2d-4203-10bd-e3f2-2c853c0c6a71.assembly 1.1" 200 2 [11.47ms]
mdadm: mdvote_log_res: /e63f1c2d-4203-10bd-e3f2-2c853c0c6a71.assembly: 14
mdadm: load_sys: /sys/block/md1/md/metadata_version: none
mdadm: sysfs_get_str: metadata_version: none

mdadm: sysfs_set_str: /sys/block/md1/md//metadata_version: external:/md0/0
mdadm: sysfs_set_str: /sys/block/md1/md//level: raid1
mdadm: sysfs_set_str: /sys/block/md1/md//raid_disks: 2
mdadm: sysfs_set_str: /sys/block/md1/md//chunk_size: 0
mdadm: sysfs_set_str: /sys/block/md1/md//layout: 0
mdadm: sysfs_set_str: /sys/block/md1/md//component_size: 98304
mdadm: sysfs_set_str: /sys/block/md1/md//resync_start: 0
mdadm: assemble_container_content: 8:32 event counter: 13 (XXX 2) is below minimum assembly sequence number: 14
mdadm: assemble_container_content: /dev/md1 assembled with 0 devices but not started
```

Can heal handle this?  If so, I think we're good!

Reply:
Hey thanks for looking!

As it turns out, you can *also* tell the difference between the two log messages because the new `assemble_container_content()` one uses the major/minor for the device name, and the previous actually uses the device name.

Yes, "assembled with 0 devices but not started" is actually exactly the right behavior.

Imagine you have `/dev/md/container` that is pre-assembled with the invalid drive `/dev/sdc`.

When you attempt to `assemble_container_content()` to `/dev/md/member-array`, it will create the dev entry (ie: `/dev/md126`), but will noticibly **NOT** create the symlink to the array named `/dev/md/member-array`. So you'll have:

```
/dev/md/container -> /dev/md127
(null)            -> /dev/md126
```

md126 is in actually in a FAILED state (visible in `mdadm --detail` output).

Heal considers this as what we call a "dangling" array. Any assembly retries, cleanup paths, etc. will "stop dangling arrays", before proceeding. System array / bitmap array code also ensures any dangling arrays are stopped before retrying. Because heal/system-array **ALWAYS** operates on the **NAMED** device entries in `/dev/md/array-name` this is a common check to ensure correct operation of the volumes.

Thanks for beating on it some from your perspective! Appreciate it.

I'll merge this in.

Reply:
Nice!  Sounds great, man!